Library/OAuth.swift

```diff
@@ -1,12 +1,52 @@
+import AuthenticationServices
 import Foundation
+import KsApi
 
 public struct OAuth {
   public init() {}
 
-  public static let redirectScheme = "ksrauth2"
-  public static func authorizationURL() -> URL {
-    let base = AppEnvironment.current.apiService.serverConfig.webBaseUrl
-    // TODO: MBL-1159: This will take URL parameters, as defined in the ticket, for PKCE
-    return base
+  static let redirectScheme = "ksrauth2"
+
+  static func authorizationURL(withCodeChallenge challenge: String) -> URL? {
+    let serverConfig = AppEnvironment.current.apiService.serverConfig
+    let baseURL = serverConfig.webBaseUrl
+
+    let parameters = [
+      URLQueryItem(name: "redirect_uri", value: redirectScheme),
+      URLQueryItem(name: "response_type", value: "code"),
+      URLQueryItem(name: "scope", value: "email"),
+      URLQueryItem(name: "client_id", value: serverConfig.apiClientAuth.clientId),
+      URLQueryItem(name: "code_challenge_method", value: "S256"),
+      URLQueryItem(name: "code_challenge", value: challenge)
+    ]
+
+    var components = URLComponents(url: baseURL, resolvingAgainstBaseURL: false)
+    components?.path = "/oauth/authorizations/new"
+    components?.queryItems = parameters
+
+    return components?.url
+  }
+
+  public static func createAuthorizationSession() -> ASWebAuthenticationSession? {
+    do {
+      let verifier = try PKCE.createCodeVerifier(byteLength: 32)
+      let challenge = try PKCE.createCodeChallenge(fromVerifier: verifier)
+      guard let url = authorizationURL(withCodeChallenge: challenge) else {
+        return nil
+      }
+
+      let session = ASWebAuthenticationSession(
+        url: url,
+        callbackURLScheme: OAuth.redirectScheme
+      ) { _, _ in
+        // TODO: MBL-1159: Exchange information in callback for credentials, then login.
+      }
+
+      return session
+
+    } catch {
+      // TODO: Is there a way we can log/monitor these errors?
```

Comment:
@scottkicks Do we have a place to log generic errors in iOS? In this case, I only expect this catch to happen if something _really odd_ occurs, i.e. we fail to create a URL or a hash.

Reply:
Added this to our iOS sync agenda.